oauthproxy.go

```diff
@@ -579,10 +579,23 @@ func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error)
 		redirect = req.Form.Get("rd")
 	}
 	if !p.IsValidRedirect(redirect) {
-		// Use RequestURI to preserve ?query
-		redirect = req.URL.RequestURI()
-		if strings.HasPrefix(redirect, p.ProxyPrefix) {
-			redirect = "/"
+		// Quirk: On reverse proxies that doesn't have support for
+		// "X-Auth-Request-Redirect" header or dynamic header/query string
+		// manipulation (like Traefik v1 and v2), we can try if the header
+		// X-Forwarded-Host exists or not.
+		if req.Host == util.GetRequestHost(req) {
```

Comment:
There is a lot of nested logic in here. We should move this to a private helper method and reduce some of the branching levels by returning `redirect` when you have it instead of all these nested if/else loops.